hal-sim/hal_impl/functions.py

```diff
@@ -856,7 +856,24 @@ def freeDIOPort(dioPortHandle):
 
 def allocateDigitalPWM(status):
     status.value = 0
-    return types.DigitalPWMHandle()
+    handle = 0
+    dio = None
+    for i, port in enumerate(hal_data["dio"]):
+        if i >= kNumDigitalHeaders:
+            if not hal_data["mxp"][i]["initialized"]:
+                dio = _initport('dio', i, status)
+                hal_data["mxp"][i]["initialized"] = True
+                handle = i
+                break
+        if not port['initialized']:
+            dio = _initport('dio', i, status)
+            handle = i
+            break
```

Comment:
I still don't understand how this matches the athena implementation.

Also, the fact that you have duplicated code smells like you're doing something terribly wrong here.